SQL: NULLIF function

SuperSQL does not yet have support for the standard SQL function [`NULLIF`](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-NULLIF).

## Details

Repro is with super commit c6ebc21. This issue was found via [a sqllogictest](https://github.com/dolthub/sqllogictest/blob/a1f4aa3c1d691552a4a03d9603c45bcf93d0bb9d/test/random/groupby/slt_good_0.test#L53154-L53159).

Here's the function being used in Postgres.

```
$ psql postgres -P "null=NULL"
psql (17.6 (Homebrew))
Type "help" for help.

postgres=# SELECT NULLIF(1,1) AS null_if_equal;
 null_if_equal 
---------------
          NULL
(1 row)

postgres=# SELECT NULLIF(1,2) AS null_if_equal;
 null_if_equal 
---------------
             1
(1 row)
```

Whereas currently in `super`:

```
$ super -version
Version: c6ebc2139

$ super -c "SELECT NULLIF(1,1) AS null_if_equal;"
no such function at line 1, column 8:
SELECT NULLIF(1,1) AS null_if_equal;
       ~~~~~~~~~~~
```

I can see this function mentioned as far back as the [1992 SQL spec](https://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt).

I've also spotted queries such as the one from [this sqllogictest](https://github.com/dolthub/sqllogictest/blob/a1f4aa3c1d691552a4a03d9603c45bcf93d0bb9d/test/random/expr/slt_good_83.test#L1088-L1092) that panic when `NULLIF` is attempted. Here's a simplified version of that query that also panics:

```
$ super -version
Version: 9db7691f8

$ super -c "SELECT COALESCE (NULLIF (2,1), 3) * COUNT (*);"
panic: (*sem.BadExpr) 0xc00002a550

goroutine 1 [running]:
github.com/brimdata/super/compiler/semantic.keySubst({0x83d68a0, 0xc00002a550}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/projection.go:252 +0xa1a
github.com/brimdata/super/compiler/semantic.keySubst({0x83d6690, 0xc000030000}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/projection.go:180 +0xe3e
github.com/brimdata/super/compiler/semantic.keySubst({0x83d6660, 0xc000030040}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/projection.go:174 +0x556
github.com/brimdata/super/compiler/semantic.(*translator).genAggregateOutput(0xc0000ff110, {0x83cb8c8, 0xc000015f50}, {0xc00001dec0?, 0xc00001de40?, 0xc00002a4e0?}, {0x0, 0x0, 0x0}, {0xc00001df40, ...})
	/Users/phil/work/super/compiler/semantic/sql.go:232 +0x358
github.com/brimdata/super/compiler/semantic.(*translator).genAggregate(0xc0000ff110, {0x0, 0x2c}, {0xc00001dec0?, 0x0?, 0x1}, {0x0, 0x0}, {0x0, 0x0, ...}, ...)
	/Users/phil/work/super/compiler/semantic/sql.go:206 +0x11d4
github.com/brimdata/super/compiler/semantic.(*translator).sqlSelect(0xc0000ff110, 0xc000408510, {0x0?, 0x0?, 0xc00001b2f0?})
	/Users/phil/work/super/compiler/semantic/sql.go:58 +0x86b
github.com/brimdata/super/compiler/semantic.(*translator).sqlQueryBody(0x0?, {0x83d4bb0?, 0xc000408510?}, {0x0?, 0x0?, 0x0?})
	/Users/phil/work/super/compiler/semantic/sql.go:451 +0x28b
github.com/brimdata/super/compiler/semantic.(*translator).sqlQueryBody(0xc0000ff110, {0x83d4b80?, 0xc000779b80}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/sql.go:459 +0x1ab
github.com/brimdata/super/compiler/semantic.(*translator).semOp(0xc0000ff110, {0x83d51e0?, 0xc000027290}, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/semantic/op.go:541 +0xc29
github.com/brimdata/super/compiler/semantic.(*translator).seq(...)
	/Users/phil/work/super/compiler/semantic/op.go:33
github.com/brimdata/super/compiler/semantic.Analyze({0x83dd3e0?, 0xc000027a70?}, 0xc000021bc0, 0x6441aa5?, 0x0)
	/Users/phil/work/super/compiler/semantic/analyzer.go:26 +0x1d0
github.com/brimdata/super/compiler.Analyze(...)
	/Users/phil/work/super/compiler/package.go:27
github.com/brimdata/super/compiler.CompileWithAST(0xc000027a70, 0x6570b65?, 0xc0000ff0a0, 0x1, 0xc, {0x0, 0x0, 0x0})
	/Users/phil/work/super/compiler/package.go:68 +0x65
github.com/brimdata/super/compiler.(*compiler).NewQuery(0x83dd488?, 0xc000136870?, 0xc0001d99a8?, {0x0?, 0x90?, 0x8358340?}, 0x649d4a0?)
	/Users/phil/work/super/compiler/compiler.go:42 +0x45
github.com/brimdata/super/runtime.CompileQuery({0x83dd488?, 0xc000136870?}, 0xc0004085a0, {0x83ca2a8, 0xc000088640}, 0xc000021bc0, {0x0, 0x0, 0x0})
	/Users/phil/work/super/runtime/compiler.go:34 +0x10f
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00012a488, {0xc00003e8f0, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:153 +0x591
github.com/brimdata/super/pkg/charm.path.run({0xc000132190, 0x1, 0x1}, {0xc00003e8f0, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0x9547380, {0xc00003e8e0, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:74 +0x1fa
main.main()
	/Users/phil/work/super/cmd/super/main.go:39 +0x5b
```

If I drop the `COUNT (*)` portion, it gives the expected error message instead.

```
$ super -c "SELECT COALESCE (NULLIF (2,1), 3);"
no such function at line 1, column 18:
SELECT COALESCE (NULLIF (2,1), 3);
                 ~~~~~~~~~~~~
```

And heres Postgres returning the expected result.

```
$ psql postgres -c "SELECT COALESCE (NULLIF (2,1), 3) * COUNT (*);"
 ?column? 
----------
        2
(1 row)
```

--- a/runtime/sam/expr/function/function.go
+++ b/runtime/sam/expr/function/function.go
@@ -115,6 +115,9 @@ func New(sctx *super.Context, name string, narg int) (expr.Function, error) {
 		argmax = 0
 		argmin = 0
 		f = &Now{}
+	case "nullif":
+		argmin, argmax = 2, 2
+		f = newNullIf()
 	case "parse_sup":
 		f = newParseSUP(sctx)
 	case "parse_uri":

--- a/runtime/sam/expr/function/nullif.go
+++ b/runtime/sam/expr/function/nullif.go
@@ -1,0 +1,29 @@
+package function
+
+import (
+	"github.com/brimdata/super"
+	"github.com/brimdata/super/order"
+	"github.com/brimdata/super/runtime/sam/expr"
+)
+
+type NullIf struct {
+	compareFn expr.CompareFn
+}
+
+func newNullIf() *NullIf {
+	return &NullIf{expr.NewValueCompareFn(order.Asc, order.NullsLast)}
+}
+
+func (n *NullIf) Call(args []super.Value) super.Value {
+	val0, val1 := args[0].Under(), args[1].Under()
+	if val0.IsError() {
+		return val0
+	}
+	if val1.IsError() {
+		return val1
+	}
+	if n.compareFn(val0, val1) == 0 {
+		return super.NewValue(val0.Type(), nil)
+	}
+	return val0
+}

--- a/runtime/vam/expr/compare.go
+++ b/runtime/vam/expr/compare.go
@@ -23,6 +23,10 @@ func NewCompare(sctx *super.Context, op string, lhs, rhs Evaluator) *Compare {
 	return &Compare{sctx, vector.CompareOpFromString(op), lhs, rhs}
 }
 
+func (c *Compare) Compare(vec0, vec1 vector.Any) vector.Any {
+	return c.eval(vec0, vec1)
+}
+
 func (c *Compare) Eval(val vector.Any) vector.Any {
 	return vector.Apply(true, c.eval, c.lhs.Eval(val), c.rhs.Eval(val))
 }

--- a/runtime/vam/expr/function/function.go
+++ b/runtime/vam/expr/function/function.go
@@ -87,6 +87,9 @@ func New(sctx *super.Context, name string, narg int) (expr.Function, error) {
 	case "network_of":
 		argmax = 2
 		f = &NetworkOf{sctx}
+	case "nullif":
+		argmin, argmax = 2, 2
+		f = newNullIf(sctx)
 	case "parse_sup":
 		f = newParseSUP(sctx)
 	case "parse_uri":

--- a/runtime/vam/expr/function/nullif.go
+++ b/runtime/vam/expr/function/nullif.go
@@ -1,0 +1,39 @@
+package function
+
+import (
+	"github.com/brimdata/super"
+	"github.com/brimdata/super/runtime/vam/expr"
+	"github.com/brimdata/super/vector"
+	"github.com/brimdata/super/vector/bitvec"
+)
+
+type NullIf struct {
+	compare *expr.Compare
+}
+
+func newNullIf(sctx *super.Context) *NullIf {
+	return &NullIf{expr.NewCompare(sctx, "==", nil, nil)}
+}
+
+func (n *NullIf) Call(vecs ...vector.Any) vector.Any {
+	if vecs[0].Type().Kind() == super.ErrorKind {
+		return vecs[0]
+	}
+	if vecs[1].Type().Kind() == super.ErrorKind {
+		return vecs[1]
+	}
+	result := n.compare.Compare(vecs[0], vecs[1])
+	if result.Type().Kind() == super.ErrorKind {
+		return vecs[0]
+	}
+	nulls := vector.NullsOf(vecs[0]).Clone()
+	for i := range result.Len() {
+		if b, _ := vector.BoolValue(result, i); b {
+			if nulls.IsZero() {
+				nulls = bitvec.NewFalse(vecs[0].Len())
+			}
+			nulls.Set(i)
+		}
+	}
+	return vector.CopyAndSetNulls(vecs[0], nulls)
+}

--- a/runtime/ztests/expr/function/nullif.yaml
+++ b/runtime/ztests/expr/function/nullif.yaml
@@ -1,0 +1,33 @@
+spq: nullif(this[0], this[1])
+
+vector: true
+
+input: |
+  [1, 1]
+  [1, 1::int32]
+  [1::int32, 1]
+  [1, 2]
+  [1, "s"]
+  [1, null]
+  [1, error(0)]
+  [null, 1]
+  [null, null]
+  [null, error(0)]
+  [error(0), 1]
+  [error(0), null]
+  [error(0), error(1)]
+
+output: |
+  null::int64
+  null::int64
+  null::int32
+  1
+  1
+  1
+  error(0)
+  null::int64
+  null
+  null::error(int64)
+  error(0)
+  error(0)
+  error(0)

--- a/vector/bitvec/bits.go
+++ b/vector/bitvec/bits.go
@@ -2,6 +2,7 @@ package bitvec
 
 import (
 	"math/bits"
+	"slices"
 	"strings"
 )
 
@@ -26,6 +27,10 @@ func NewTrue(n uint32) Bits {
 		b.bits[i] = ^uint64(0)
 	}
 	return b
+}
+
+func (b Bits) Clone() Bits {
+	return Bits{slices.Clone(b.bits), b.length}
 }
 
 func (b Bits) IsZero() bool {

--- a/vector/bool.go
+++ b/vector/bool.go
@@ -1,6 +1,8 @@
 package vector
 
 import (
+	"fmt"
+
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/scode"
 	"github.com/brimdata/super/vector/bitvec"
@@ -81,7 +83,7 @@ func BoolValue(vec Any, slot uint32) (bool, bool) {
 	case *View:
 		return BoolValue(vec.Any, vec.Index[slot])
 	}
-	panic(vec)
+	panic(fmt.Sprintf("%#v", vec))
 }
 
 func NullsOf(v Any) bitvec.Bits {
@@ -211,6 +213,14 @@ func CopyAndSetNulls(v Any, nulls bitvec.Bits) Any {
 		return &copy
 	case *Union:
 		return NewUnion(v.Typ, v.Tags, v.Values, nulls)
+	case *View:
+		newNulls := bitvec.NewFalse(uint32(len(v.Index)))
+		for i, idx := range v.Index {
+			if nulls.IsSet(uint32(i)) {
+				newNulls.Set(idx)
+			}
+		}
+		return NewView(CopyAndSetNulls(v.Any, newNulls), v.Index)
 	default:
 		panic(v)
 	}

